### PR TITLE
NTOC tag

### PR DESCRIPTION
**Pull request by aartaka**
# Description

This adds a table of contents tag that uses its own body to generate the TOC.

Based on the interactive tags branch, for no particular reason.

I'll merge it after the interactive tags are merged.

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
  - [X] (If this changes something about the features showcased on Nyxt website) I have these changes described in the new/existing article at Nyxt website or will notify one of maintainters to do so.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by aartaka** [comment]
> > Based on the interactive tags branch, for no particular reason.
> 
> @aartaka, I'm lost here! Are you referring to PR #2842?

Yes, that one.

**Comment by aartaka** [comment]
Okay, rebased over master so that `:ntoc` can be judged for its virtues alone.

**Comment by aadcg** [comment]
Thanks. I think PRs should always be rebased against master unless there's a very particular reason otherwise.

I'll look at this later.

File: source/spinneret-tags.lisp [hunk 1]
```diff
@@ -534,3 +534,42 @@ Calls Lisp forms in ONFOCUS and ONCHANGE when one focuses and edits the input (r
...
         ,@attrs
         (:raw (the string ,input-contents))))))
+
+(serapeum:-> %ntoc-create-toc (string) *)
+(defun %ntoc-create-toc (body)
+  "Generate the code for the table of contents based on string BODY."
+  (labels ((parent-section (elem)
+             (find-if #'nyxt/dom:section-element-p (nyxt/dom:parents elem)))
+           (fm (control &rest args)
+             (apply #'format nil control args))
```
Comment (Ambrevar) [inline source/spinneret-tags.lisp:544]: Seems a bit unnecessary.  Also there is `sera:fmt`.

File: source/spinneret-tags.lisp [hunk 2]
```diff
@@ -534,3 +534,42 @@ Calls Lisp forms in ONFOCUS and ONCHANGE when one focuses and edits the input (r
...
+           (format-section (heading level)
+             (with-html-string
+               (let ((parent-section (parent-section heading)))
+                 (:li (:a :href (fm "#~a" (plump:attribute parent-section "id"))
+                          (plump:text heading)))
+                 (serapeum:and-let* ((_ (< level 6))
```
Comment (Ambrevar) [inline source/spinneret-tags.lisp:550]: Make this `6` an option?

**Comment by aartaka** [comment]
I've merged it, because 
- it seems to be the best shot at making a TOC tag that I can possibly create, 
- CI passes,
- and all of @Ambrevar comments are addressed.

**Comment by aartaka** [comment]
>     * CI passes,

Except it doesn't :D Fixed on master.

**Comment by Ambrevar** [comment]
Thanks!

**Pull request merged**